skimage/transform/_seam_carving.pyx

```diff
@@ -0,0 +1,226 @@
+                            Py_ssize_t cols) nogil:
+    """ For each row, compute the lowest seam value for all its columns.
+
+    This function updates `cumulative_img` such that `cumulative_img[r, c]`
+    is the total energy of the lowest energy seam ending at `(r, c)`.
+
+    Parameters
+    ----------
+    energy_img : (M, N, 1) ndarray
+        Cost array representing the expense to remove each pixel. Seam carving
+        tries to avoid pixels with high costs.
+    cumulative_img : (M, N) ndarray
+        The array to be updated inplace with the total cost of lowest energy
+        seams.
+    track_img : (M, N) ndarray
+        For each pixel, `track_img` stores the relative column offset in
+        the previous row which has the lowest value in `cumulative_img`. This
+        helps in in re-tracing the minimum cost seam.
+    cols : int
+        Number of columns to process.
+    """
+
+    cdef Py_ssize_t r, c, offset, c_idx
+    cdef Py_ssize_t rows = energy_img.shape[0]
+    cdef cnp.double_t min_cost = DBL_MAX
+    cdef Py_ssize_t colsm1 = cols - 1
+    cdef Py_ssize_t rm1
+
+    for c in range(cols):
+        cumulative_img[0, c] = energy_img[0, c, 0]
+
+    for r in range(1, rows):
+        rm1 = r - 1
+        for c in range(cols):
+            min_cost = DBL_MAX
+            for offset in range(-1, 2):
+
+                c_idx = c + offset
+                if (c_idx > colsm1) or (c_idx < 0):
+                    continue
+
+                if cumulative_img[rm1, c_idx] < min_cost:
+                    min_cost = cumulative_img[rm1, c_idx]
+                    track_img[r, c] = offset
+
+            cumulative_img[r, c] = min_cost + energy_img[r, c, 0]
+
+cdef cnp.uint8_t _mark_seam(cnp.int8_t[:, ::1] track_img,
+                            Py_ssize_t start_index,
+                            cnp.uint8_t[:, ::1] seam_map,
+                            Py_ssize_t[::1] seam_buffer) nogil:
+
+    """ Re-trace the optimal seam from a given column in the last row.
+
+    This function tries to re-track an optimal seam from `start_index` and
+    tries to mark it in `seam_map`. If this seam intersects with any existing
+    seam in `seam_map` the function returns `0` without marking anything. Else
+    it marks the seam in `seam_map` and returns `1`.
+
+    track_img : (M, N) ndarray
+       The array of relative column indices as updated by `_preprocess_image`.
+    start_index : int
+       The column number of the bottom most row from where to start re-tracing
+       the seam.
+    seam_map : (M, N) ndarray
+        The array used to mark seams. If a pixel is marked as as seam it is set
+        to `1`, else `0`.
+    seam_buffer : (M,) ndarray
+        Buffer used to store the column indices of the seam currently being
+        checked. This is preallocated to save time.
+
+    Returns
+    -------
+    out : int
+        `1` if seam was marked, `0` is seam inersects and was not marked.
```

Comment:
don't call it `out`, call it `success`. Perhaps it's worth returning a bool? Finally, "`0` **if** seam intersects"